src/Domain/CoordinateMaps/TimeDependent/ShapeMapTransitionFunctions/Wedge.cpp

```diff
@@ -0,0 +1,484 @@
+}
+
+template <typename T>
+T Wedge::map_over_radius_impl(const std::array<T, 3>& source_coords) const {
+  const std::array<T, 3> rotated_coords =
+      discrete_rotation(orientation_map_.inverse_map(), source_coords);
+  check_distances(rotated_coords);
+  const T radius = magnitude(rotated_coords);
+  const T outer_distance = overall_outer_surface_.distance(rotated_coords);
+
+  return (outer_distance - radius) /
+         ((outer_distance - overall_inner_surface_.distance(rotated_coords)) *
+          radius);
+}
+
+template <typename T>
+std::array<T, 3> Wedge::gradient_impl(
+    const std::array<T, 3>& source_coords) const {
+  // If both surfaces are spherical then we short circuit because the distances
+  // are constant and we only need to take a derivative of r.
+  // (grad f)_i = -(x_i/r)/(D_out - D_in)
+  const std::array<T, 3> rotated_coords =
+      discrete_rotation(orientation_map_.inverse_map(), source_coords);
+  check_distances(rotated_coords);
+  if (overall_inner_surface_.sphericity == 1.0 and
+      overall_outer_surface_.sphericity == 1.0) {
+    const T one_over_denom =
+        1.0 / (magnitude(rotated_coords) *
+               (overall_outer_surface_.distance(rotated_coords) -
+                overall_inner_surface_.distance(rotated_coords)));
+
+    return discrete_rotation(orientation_map_,
+                             -rotated_coords * one_over_denom);
+  }
+
+  const T radius = magnitude(rotated_coords);
+
+  const T one_over_radius = 1.0 / radius;
+  T outer_distance = overall_outer_surface_.distance(rotated_coords);
+  const T one_over_denom =
+      1.0 / (outer_distance - overall_inner_surface_.distance(rotated_coords));
+  T& outer_distance_minus_radius = outer_distance;
+  outer_distance_minus_radius = outer_distance - radius;
+
+  // Avoid small negative numbers if we are at outer boundary
+  for (size_t i = 0; i < get_size(radius); i++) {
+    if (equal_within_roundoff(get_element(outer_distance_minus_radius, i),
+                              0.0)) {
+      get_element(outer_distance_minus_radius, i) = 0.0;
+    }
+  }
+
+  // Regardless of the sphericities below, we always need this factor in the
+  // first term so we calculate it now.
+  std::array<T, 3> result = -1.0 * rotated_coords * one_over_radius;
+
+  const auto make_factor = [&one_over_radius](const Surface& surface) -> T {
+    return (1.0 - surface.sphericity) * surface.radius * cube(one_over_radius) /
+           sqrt(3.0);
+  };
+
+  // We can make some simplifications if either of the surfaces are spherical
+  // because then the derivative of the distance is zero since it's constant. In
+  // the first two branches, it's safe to assume the other sphericity isn't 1
+  // because of the above check.
+  if (overall_outer_surface_.sphericity == 1.0) {
+    const T total_factor = one_over_denom * outer_distance_minus_radius *
+                           make_factor(overall_inner_surface_);
```

Comment:
Since you seem concerned with such things in this function, pretty sure `T total_factor = make_factor(...); total_factor *= ...;` saves you an allocation.